snuba/query/snql/__init__.py

```diff
@@ -1,23 +1,280 @@
-    match_clause = space* "MATCH" "(" clause ")" space*
-    where_clause = space* "WHERE" clause space*
-    collect_clause = space* "COLLECT" clause "BY" clause space*
-    having_clause = space* "HAVING" clause space*
-    order_by_clause = space* "ORDER BY" clause ("ASC"/"DESC") space*
+    low_pri_arithmetic    = space* high_pri_arithmetic space* (low_pri_tuple)*
+    high_pri_arithmetic   = space* arithmetic_term space* (high_pri_tuple)*
+    low_pri_tuple         = low_pri_op high_pri_arithmetic
+    high_pri_tuple        = high_pri_op arithmetic_term
+    arithmetic_term       = (space*) (function_call / numeric_literal / column_name) (space*)
+    low_pri_op            = "+" / "-"
+    high_pri_op           = "/" / "*"
+    param_expression      = low_pri_arithmetic / quoted_literal
+    parameters_list       = parameter* (param_expression)
+    parameter             = param_expression space* comma space*
+    function_call         = function_name open_paren parameters_list? close_paren (open_paren parameters_list? close_paren)?
+    simple_term           = quoted_literal / numeric_literal / column_name
+    literal               = ~r"[a-zA-Z0-9_\.:-]+"
+    quoted_literal        = "'" string_literal "'"
+    string_literal        = ~r"[a-zA-Z0-9_\.\+\*\/:-]*"
+    numeric_literal       = ~r"-?[0-9]+(\.[0-9]+)?(e[\+\-][0-9]+)?"
+    column_name           = ~r"[a-zA-Z_][a-zA-Z0-9_\.]*"
+    function_name         = ~r"[a-zA-Z_][a-zA-Z0-9_]*"
+    open_paren            = "("
+    close_paren           = ")"
+    space                 = " "
+    comma                 = ","
 
-    clause = space* ~r"[-=><\w]+" space*
-    space = " "
 """
 )
 
 
+class LowPriOperator(Enum):
+    PLUS = "+"
+    MINUS = "-"
+
+
+class HighPriOperator(Enum):
+    MULTIPLY = "*"
+    DIVIDE = "/"
+
+
+class LowPriTuple(NamedTuple):
+    op: LowPriOperator
+    arithm: Expression
+
+
+class HighPriTuple(NamedTuple):
+    op: HighPriOperator
+    arithm: Expression
+
+
+HighPriArithmetic = Union[Node, HighPriTuple, Sequence[HighPriTuple]]
+LowPriArithmetic = Union[Node, LowPriTuple, Sequence[LowPriTuple]]
+
+
+def get_arithmetic_function(
+    operator: Enum,
+) -> Callable[[Expression, Expression, Optional[str]], FunctionCall]:
+    return {
+        HighPriOperator.MULTIPLY: multiply,
+        HighPriOperator.DIVIDE: divide,
+        LowPriOperator.PLUS: plus,
+        LowPriOperator.MINUS: minus,
+    }[operator]
+
+
+def get_arithmetic_expression(
+    term: Expression, exp: Union[LowPriArithmetic, HighPriArithmetic],
+) -> Expression:
+    if isinstance(exp, Node):
+        return term
+    if isinstance(exp, (LowPriTuple, HighPriTuple)):
+        return get_arithmetic_function(exp.op)(term, exp.arithm, None)
+    elif isinstance(exp, list):
+        for elem in exp:
+            term = get_arithmetic_function(elem.op)(term, elem.arithm, None)
+    return term
+
+
+class SnQLVisitor(NodeVisitor):
```

Comment:
You are right, there is a lot of duplication between here and the old expression visitor, I would advise to move that to its own class:
1) create an ExpressionVisitor (do not inherit NodeVisitor), it just contains the methods that produce the expressions
2) move all the common visit_* methods and the classes aruond them into this class.
3) reuse such class from both the visit_methods of this one and the old clickhouse visitor by delegating calls from the visit methods in this class to this external visitor.

I would advise against making this visitor depend directly on the old one. You would have to manage the instance of such visitor, which is something you cannot do reliable (it extends NodeVisitor and the state is handled by Parsimonious).